### PR TITLE
refactor: cloudinary driver for goravel

### LINKED ISSUE #1: 🐛 [Bug] Some Suggestions
### Before feedback (在反馈之前)

- [X] I've searched the existing Issues, Discussions and Google (我已经搜索了现有的 Issues, Discussions 和 Google)
- [X] The problem can be stably reproduced (这个问题可以被稳定复现)
- [X] The problem is generated after upgrading (问题是在升级之后产生的)

### Operating System (操作系统)

Linux

### Golang Version (Go 版本)

1.18.x

### Goravel Version (Goravel 版本)

1.12.6

### Describe The Problem (描述问题)

1. Optimize README

image

1.1 `cloudinaryFacades "github.com/goravel/cloudinary"` -> `cloudinaryfacades "github.com/goravel/cloudinary/facades"`
1.2 Optimize configuration like other storage packages;

image

2. The `config/cloudinary.go` and `contracts/cloudinary.go` files seem unnecessary;
3. Please add the LICENSE file;
4. Order import for the `cloudinary.go` file;
5. I think `AllDirectories` and `getAllDirectoriesRecursively` can be merged, `AllDirectories` is very simple, it's unnecessary to separate them;
6. [This line](https://github.com/goravel/cloudinary/blob/b7c0c09cfd137a1d1230e5748723766350bf1ab1/cloudinary.go#L69) is so strange, I suggest optimize it;
7. Does `AssetType` must be set in [this line](https://github.com/goravel/cloudinary/blob/b7c0c09cfd137a1d1230e5748723766350bf1ab1/cloudinary.go#L83C2-L83C63)?  Are there other `AssetType`?
8. Could you please reset the repository and create a PR for the current files? I found there are many places that need to be optimized, it's hard to record them here.

### Reproduction Code (复现代码或截图)

No

### PR DESCRIPTION
**Pull request by kkumar-gcc**
Closes #1 

## 📑 Description
This PR will remove unnecessary folders and clean up code-base for  review.

## ✅ Checks

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [x] All the tests have passed

### CODE REVIEW

File: cloudinary_test.go [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+	assert.Nil(t, os.WriteFile("test.txt", []byte("Goravel"), 0644))
+
+	mockConfig := &configmocks.Config{}
+	mockConfig.On("GetString", "filesystems.disks.cloudinary.key").Return(os.Getenv("CLOUDINARY_API_KEY"))
+	mockConfig.On("GetString", "filesystems.disks.cloudinary.secret").Return(os.Getenv("CLOUDINARY_API_SECRET"))
+	mockConfig.On("GetString", "filesystems.disks.cloudinary.cloud").Return(os.Getenv("CLOUDINARY_CLOUD_NAME"))
+	mockConfig.On("GetString", "filesystems.disks.cloudinary.url").Return(os.Getenv("CLOUDINARY_URL"))
+
+	var driver contractsfilesystem.Driver
+	//url := os.Getenv("CLOUDINARY_URL")
```
Comment (hwbrzzl) [inline cloudinary_test.go:32]:
```suggestion
```

File: cloudinary_test.go [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+		//		assert.Equal(t, url, driver.Url("Url/1.txt"))
+		//		resp, err := http.Get(url)
+		//		assert.Nil(t, err)
+		//		content, err := io.ReadAll(resp.Body)
+		//		assert.Nil(t, resp.Body.Close())
+		//		assert.Nil(t, err)
+		//		assert.Equal(t, "Goravel", string(content))
+		//		assert.Nil(t, driver.DeleteDirectory("Url"))
+		//	},
+		//},
```
Comment (hwbrzzl) [inline cloudinary_test.go:357]: Q: Why remove this?

File: utils.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+func validPath(path string) string {
+	realPath := strings.TrimPrefix(path, "."+string(filepath.Separator))
+	realPath = strings.TrimPrefix(realPath, string(filepath.Separator))
+	realPath = strings.TrimPrefix(realPath, ".")
+	realPath = strings.TrimSuffix(realPath, string(filepath.Separator))
+	//if realPath != "" && !strings.HasSuffix(realPath, string(filepath.Separator)) {
+	//	realPath += string(filepath.Separator)
+	//}
```
Comment (hwbrzzl) [inline utils.go:87]: S: Remove this if not needed.

**Review by hwbrzzl — DISMISSED** [review]
Great PR! A few questions.

File: go.mod [hunk 4]
```diff
@@ -0,0 +1,30 @@
+module github.com/goravel/cloudinary
+
+go 1.19
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply go.mod:3]: It also supports Go 1.19. I will update it.

File: service_provider.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+	app.BindWith(Binding, func(app foundation.Application, parameters map[string]any) (any, error) {
+		return NewCloudinary(context.Background(), app.MakeConfig(), parameters["disk"].(string))
+	})
+}
+func (receiver *ServiceProvider) Boot(app foundation.Application) {
+	app.Publishes("github.com/goravel/cloudinary", map[string]string{
+		"config/cloudinary.go": app.ConfigPath(""),
+	})
+}
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply service_provider.go:26]: Then where should move it ?

File: utils.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	"fmt"
+	"github.com/cloudinary/cloudinary-go/v2/api"
+	"github.com/goravel/framework/contracts/filesystem"
+	"github.com/goravel/framework/support/file"
+	"io"
+	"net/http"
+	"path"
+	"path/filepath"
+	"strings"
+)
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply utils.go:13]: Thanks for suggestion. I have updated it.

File: cloudinary.go [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+	result, err := r.instance.Upload.Upload(r.ctx, resource.SecureURL, uploader.UploadParams{
+		PublicID: newFile,
+	})
+	if err != nil {
+		return err
+	}
+	// Check if the public_id matches the newFile
+	if result.PublicID != newFile {
+		return fmt.Errorf("copy file error: public_id mismatch, expected %s but got %s", newFile, result.PublicID)
+	}
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply cloudinary.go:108]: Yeah, it does have an `Error` interface.

**Comment by hwbrzzl** [comment]
Please add the `.github` folder like this: https://github.com/goravel/redis/tree/master/.github

File: cloudinary.go [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+	explicitResult, err := r.instance.Upload.Explicit(r.ctx, uploader.ExplicitParams{
+		PublicID:     file,
+		RawConvert:   file,
+		ResourceType: string(assetType),
+		Type:         "upload",
+	})
+	if err != nil {
+		return "", err
+	}
+	if explicitResult.Error.Message != "" {
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply cloudinary.go:211]: `explicitResult.Error` is not interface so i can't judge it like this.
```go
// Type of explicitResult.Error is api.ErrorResp
// ErrorResp is the failed api request main struct.
type ErrorResp struct {
	Message string `json:"message"`
}
```

File: cloudinary.go [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+
+// AllFiles returns all the files from the given directory including all its subdirectories.
+func (r *Cloudinary) AllFiles(path string) ([]string, error) {
+	var result []string
+	assetTypes := []api.AssetType{api.Image, api.Video, api.File}
+	for _, assetType := range assetTypes {
+		folders, err := r.instance.Admin.Assets(r.ctx, admin.AssetsParams{
+			Prefix:       validPath(path),
+			DeliveryType: "upload",
+			AssetType:    assetType,
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply cloudinary.go:80]: As comment suggest it can only be used when you're downloading folders with all asset.(I tried, but it didn't work. )

File: utils.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+	fileExtension := strings.ToLower(filepath.Ext(fileName))
+	// Check if the file has an extension or not.
+	if fileExtension == "" {
+		// Assuming "Image" asset type for files with no extension.
+		return api.Image
+	}
+
+	switch fileExtension {
+	case ".ai", ".avif", ".bmp", ".bw", ".djvu", ".dng", ".ps", ".ept", ".eps", ".eps3", ".fbx", ".flif", ".gif", ".glb", ".heif", ".heic", ".ico", ".indd", ".jpg", ".jpe", ".jpeg", ".jp2", ".wdp", ".jxr", ".hdp", ".jxl", ".obj", ".pdf", ".ply", ".png", ".psd", ".arw", ".cr2", ".svg", ".tga", ".tif", ".tiff", ".u3ma", ".usdz", ".webp":
+		RemoveFileExtension(file)
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply utils.go:36]: Wherever we are searching for any file, We are expecting that user will use public_id() which doesn't include a file extension for videos and image asset_type) but if user includes file extension by mistake, we will remove (other than raw[File] type).

File: utils.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+		return api.Image
+	case ".3g2", ".3gp", ".avi", ".flv", ".m3u8", ".ts", ".m2ts", ".mts", ".mov", ".mkv", ".mp4", ".mpeg", ".mpd", ".mxf", ".ogv", ".webm", ".wmv", ".aac", ".aiff", ".amr", ".flac", ".m4a", ".mp3", ".ogg", ".opus", ".wav":
+		RemoveFileExtension(file)
+		return api.Video
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply utils.go:40]: I have included all types mentioned in cloudinary docs here if there is a new type it will be considered as FIle by default (raw). [docs](https://cloudinary.com/documentation/image_transformations#image_format_support) I will look a better way if there is.

File: cloudinary_test.go (hunk 2 above)
Comment (kkumar-gcc) [reply cloudinary_test.go:357]: It's temporary. It will have different url every time, so it was hard to find (it wasn't predictable I will replace it later.)

File: cloudinary.go [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+
+// Put stores a new file on the disk.
+func (r *Cloudinary) Put(file, content string) error {
+	tempFile, err := r.tempFile(content)
+	defer os.Remove(tempFile.Name())
+	if err != nil {
+		return err
+	}
+	_, err = r.instance.Upload.Upload(r.ctx, tempFile.Name(), uploader.UploadParams{
+		PublicID:    file,
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply cloudinary.go:300]: we don't it will automatically take care of it

File: cloudinary.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (hwbrzzl) [reply cloudinary.go:80]: Can we use `Auto` instead of cycle `Image`, `Video` and `File`?

File: service_provider.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (hwbrzzl) [reply service_provider.go:26]: Sorry, *remove

File: utils.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (hwbrzzl) [reply utils.go]: We must be missing some types when `cloudinary` adds new types  in the future ~

File: utils.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (hwbrzzl) [reply utils.go]: What does `public_id` look like? Can't `public_id` contains extension?

File: cloudinary.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (hwbrzzl) [reply cloudinary.go:211]: Make sense

File: cloudinary_test.go (hunk 2 above)
Comment (hwbrzzl) [reply cloudinary_test.go:357]: For example for the URL?

File: utils.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply utils.go]: Okay i have solved now we don't have to remove file extension.

File: README.md [hunk 13]
```diff
@@ -21,24 +21,31 @@ A cloudinary driver for facades.Storage of Goravel.
...
+            "cloudinary": map[string]any{
+                  "driver": "custom",
+                  "cloud":  config.Env("CLOUDINARY_CLOUD"),
+                  "key":    config.Env("CLOUDINARY_ACCESS_KEY_ID"), 
+                  "secret": config.Env("CLOUDINARY_ACCESS_KEY_SECRET"),
+                  "resource_types": map[string][]string{
+                       "image": {"png"},
+                       "video": {},
+                       "raw":   {"txt", "pdf"},
+                  },
```
Comment (hwbrzzl) [inline README.md:45]: Q: Why is this needed?

File: cloudinary.go [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+	return result, nil
+}
+
+// Exists checks if a file exists in the Cloudinary storage.
+func (r *Cloudinary) Exists(file string) bool {
+	asset, err := r.getAsset(file)
+	if asset.Error.Message != "" {
+		return false
+	}
+	return err == nil
```
Comment (hwbrzzl) [inline cloudinary.go:178]: S: The error needs to be judged first.

File: cloudinary.go [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+	resource, err := r.getAsset(file)
+	if err != nil {
+		return 0, err
+	}
+	return int64(resource.Bytes), nil
+}
+
+// TemporaryUrl get the temporary url of a file.
+func (r *Cloudinary) TemporaryUrl(file string, time time.Time) (string, error) {
+	return "", errors.New("we don't support temporary url for cloudinary")
```
Comment (hwbrzzl) [inline cloudinary.go:319]:
```suggestion
	return "", errors.New("Cloudinary doesn't support temporary url")
```

File: cloudinary.go [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+				value = resourceType
+				break
+			}
+		}
+	}
+
+	return value
+}
+
+func (r *Cloudinary) getPublicId(path string) string {
```
Comment (hwbrzzl) [inline cloudinary.go:378]: Q: Can't PublicId contain extension?

Comment (kkumar-gcc) [reply cloudinary.go:378]: No, PublicId doesn't contain extension other than `raw` type.

File: cloudinary.go (hunk 14 above)
Comment (kkumar-gcc) [reply cloudinary.go:178]: I guess we don't need to judge it since we are not sending an error anyway. We just want to check if there is an error. If yes, we will return false; else, true.

File: README.md (hunk 13 above)
Comment (kkumar-gcc) [reply README.md:45]: Instead of supporting all types, let's allow users to decide which types they want to support and in which asset_type.

File: cloudinary.go (hunk 14 above)
Comment (hwbrzzl) [reply cloudinary.go:178]: `assert` is a point type, it may be `nil` when `err != nil`, so we need to judge `err != nil` first.

File: README.md (hunk 13 above)
Comment (hwbrzzl) [reply README.md:45]: Okay, it's a good feature, I think we can set the configuration as optional, accept all types when the configuration doesn't set.

File: cloudinary.go (hunk 16 above)
Comment (hwbrzzl) [reply cloudinary.go:378]: What's the PublicId looks like?

Comment (kkumar-gcc) [reply cloudinary.go:378]: > - The Public ID value for images and videos should not include a file extension. Include the file extension for raw files only.
> - Can be up to 255 characters, including non-English characters, periods (.), forward slashes (/), underscores (_), hyphens (-).
> -  Public ID values cannot begin or end with a space or forward slash (/). Additionally, they cannot include the following characters: ? & # \ % < > +

ref: https://support.cloudinary.com/hc/en-us/community/posts/360000303551/comments/7949396230802

Comment (hwbrzzl) [reply cloudinary.go:378]: Okay, thanks!

File: README.md (hunk 13 above)
Comment (kkumar-gcc) [reply README.md:45]: Okay

File: README.md [hunk 17]
```diff
@@ -21,24 +21,31 @@ A cloudinary driver for facades.Storage of Goravel.
...
+      )
+   
+      "disks": map[string]filesystem.Disk{
+            ...
+            "cloudinary": map[string]any{
+                  "driver": "custom",
+                  "cloud":  config.Env("CLOUDINARY_CLOUD"),
+                  "key":    config.Env("CLOUDINARY_ACCESS_KEY_ID"), 
+                  "secret": config.Env("CLOUDINARY_ACCESS_KEY_SECRET"),
+                  "resource_types": map[string][]string{
```
Comment (hwbrzzl) [inline README.md:41]: S: Could you please add a comment for this? To remind the configuration is optional and its action.

File: cloudinary.go [hunk 18]
```diff
@@ -0,0 +1,393 @@
...
+
+// Exists checks if a file exists in the Cloudinary storage.
+func (r *Cloudinary) Exists(file string) bool {
+	asset, err := r.getAsset(file)
+	if err != nil {
+		return false
+	}
+	if asset.Error.Message != "" {
+		return false
+	}
```
Comment (hwbrzzl) [inline cloudinary.go:180]:
```suggestion
	if err != nil || asset.Error.Message != "" {
		return false
	}
```

File: utils.go [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+
+func validPath(path string) string {
+	realPath := strings.TrimPrefix(path, "."+string(filepath.Separator))
+	realPath = strings.TrimPrefix(realPath, string(filepath.Separator))
+	realPath = strings.TrimPrefix(realPath, ".")
+	realPath = strings.TrimSuffix(realPath, string(filepath.Separator))
+	return realPath
+}
+
+func defaultResourcesTypes() map[string][]string {
```
Comment (hwbrzzl) [inline utils.go:38]: Q: Are we must set this? Can't use `all`?

Comment (kkumar-gcc) [reply utils.go:38]: no we can't,i have asked on their community form.
https://community.cloudinary.com/discussion/297/how-to-empty-a-folder-using-cloudinary-go-sdk#latest

File: cloudinary.go [hunk 20]
```diff
@@ -0,0 +1,393 @@
...
+
+	if _, err := tempFile.WriteString(content); err != nil {
+		return nil, err
+	}
+
+	return tempFile, nil
+}
+
+func (r *Cloudinary) getResourceType(path string) string {
```
Comment (hwbrzzl) [inline cloudinary.go:364]: Q: It may not be accurate to get ResourceType through the extension of a file(User can change it easily), is this the way of official recommendation? Can we get the ResourceType by an official endpoint?

File: cloudinary.go [hunk 21]
```diff
@@ -0,0 +1,390 @@
...
+	tempFile, err := r.tempFile(content)
+	defer os.Remove(tempFile.Name())
+	if err != nil {
+		return err
+	}
+	_, err = r.instance.Upload.Upload(r.ctx, tempFile.Name(), uploader.UploadParams{
+		PublicID:       r.getPublicId(file),
+		UseFilename:    api.Bool(true),
+		UniqueFilename: api.Bool(false),
+		ResourceType:   r.getResourceType(file),
```
Comment (hwbrzzl) [inline cloudinary.go:284]: Maybe `ResourceType` is not required, see the official example: https://github.com/cloudinary/cloudinary-go/blob/bb1164bac6075db30220761632d2b4e456c7f629/cloudinary_test.go#L47C1-L47C1

Please check that other `ResourceType` is also not required.

File: cloudinary.go [hunk 22]
```diff
@@ -0,0 +1,391 @@
...
+		return nil, err
+	}
+
+	return tempFile, nil
+}
+
+func (r *Cloudinary) getResourceType(path string) string {
+	extension := strings.TrimPrefix(filepath.Ext(path), ".")
+	value := "image"
+	resourceTypes := r.config.Get(fmt.Sprintf("filesystems.disks.%s.resource_types", r.disk), defaultResourcesTypes()).(map[string][]string)
```
Comment (hwbrzzl) [inline cloudinary.go:365]: There seems to be no change here. Based on the below image or range the ResourceTypes. What's your choice?

telegram-cloud-photo-size-5-6246531548778182330-y

File: cloudinary.go [hunk 23]
```diff
@@ -339,11 +355,22 @@ func (r *Cloudinary) Url(file string) string {
 }
 
 func (r *Cloudinary) getAsset(path string) (*uploader.ExplicitResult, error) {
-	return r.instance.Upload.Explicit(r.ctx, uploader.ExplicitParams{
-		PublicID:     r.getPublicId(path),
-		Type:         "upload",
-		ResourceType: r.getResourceType(path),
-	})
+	// TODO: Search if there is a better way to get asset info
+	assetTypes := []api.AssetType{api.Image, api.Video, api.File}
```
Comment (hwbrzzl) [inline cloudinary.go:356]: Q: Why is there no `api.Raw`?

File: cloudinary.go [hunk 24]
```diff
@@ -278,24 +286,32 @@ func (r *Cloudinary) Put(file, content string) error {
...
+	}
+	return uploadResult.PublicID, nil
 }
 
 // PutFileAs stores a new file on the disk.
 func (r *Cloudinary) PutFileAs(path string, source filesystem.File, name string) (string, error) {
 	uploadResult, err := r.instance.Upload.Upload(r.ctx, source.File(), uploader.UploadParams{
 		Folder:         validPath(path),
-		PublicID:       r.getPublicId(name),
+		PublicID:       name,
```
Comment (hwbrzzl) [inline cloudinary.go:311]: Q: Can the name contain extension?

File: cloudinary.go (hunk 23 above)
Comment (kkumar-gcc) [reply cloudinary.go:356]: File is actually api.Raw

Comment (hwbrzzl) [reply cloudinary.go:356]: Why?

Comment (kkumar-gcc) [reply cloudinary.go:356]: > Why?

They have defined it this way:)

File: cloudinary.go (hunk 24 above)
Comment (hwbrzzl) [reply cloudinary.go:311]: I think we'd better follow the cloudinary suggestion.

Comment (kkumar-gcc) [reply cloudinary.go:311]: > I think we'd better follow the cloudinary suggestion.

Yeah, we are following cloudinay.

File: go.mod [hunk 25]
```diff
@@ -0,0 +1,30 @@
+module github.com/goravel/cloudinary
+
+go 1.18
```
Comment (hwbrzzl) [inline go.mod:3]:
```suggestion
go 1.20
```

File: .github/workflows/test.yml [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+  pull_request:
+env:
+  CLOUDINARY_ACCESS_KEY_ID: ${{ secrets.CLOUDINARY_ACCESS_KEY_ID }}
+  CLOUDINARY_ACCESS_KEY_SECRET: ${{ secrets.CLOUDINARY_ACCESS_KEY_SECRET }}
+  CLOUDINARY_CLOUD: ${{ secrets.CLOUDINARY_CLOUD }}
+jobs:
+  test:
+    strategy:
+      matrix:
+        go: [ '1.18' ]
```
Comment (hwbrzzl) [inline .github/workflows/test.yml:15]:
```suggestion
        go: [ '1.20' ]
```

**Review by hwbrzzl — COMMENTED** [review]
Great job! I think we can go ahead after these two optimizations.

**Review by hwbrzzl — APPROVED** [review]
LGTM
